### PR TITLE
adding a test for output of TL element

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**Description**
a simple test in cpp to verify that the output of strain and stress is ok for the TL element

Please mark the PR with appropriate tags: 
- FastPR

**Changelog**

### CODE REVIEW

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            std::vector<Vector> output_strains(1);
+            p_element->CalculateOnIntegrationPoints(GREEN_LAGRANGE_STRAIN_VECTOR,output_strains, r_model_part.GetProcessInfo());
+
+            std::vector<Vector> output_stress(1);
+            p_element->CalculateOnIntegrationPoints(PK2_STRESS_VECTOR,output_stress, r_model_part.GetProcessInfo());
+            
+            for(unsigned int i=0; i<reference_strain.size(); ++i)
+            {
+                KRATOS_CHECK_DOUBLE_EQUAL(output_strains[0][i], reference_strain[i]);
+                KRATOS_CHECK_DOUBLE_EQUAL(output_stress[0][i], reference_stress[i]);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]: I think here you can use the [vector check macro](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/checks.h#L58-L103)

**Comment by loumalouomega** [comment]
> **Description**
> a simple test in cpp to verify that the output of strain and stress is ok for the TL element
> 
> Please mark the PR with appropriate tags:
> 
>     * FastPR
> 
> 
> **Changelog**

But we have the python tests, this is basically the same, I am Ok with it, but I don't understand why

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            reference_stress[1] = c2 * reference_strain[0] + c1 * reference_strain[1];
+            reference_stress[2] = c3 * reference_strain[2];
+
+            std::vector<Vector> output_strains(1);
+            p_element->CalculateOnIntegrationPoints(GREEN_LAGRANGE_STRAIN_VECTOR,output_strains, r_model_part.GetProcessInfo());
+
+            std::vector<Vector> output_stress(1);
+            p_element->CalculateOnIntegrationPoints(PK2_STRESS_VECTOR,output_stress, r_model_part.GetProcessInfo());
+            
+            for(unsigned int i=0; i<reference_strain.size(); ++i)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
            for(std::size_t i=0; i<reference_strain.size(); ++i) {
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            p_element->CalculateLocalSystem(lhs,rhs,const_procinfo_ref);
+            p_element->FinalizeNonLinearIteration(const_procinfo_ref);
+            p_element->FinalizeSolutionStep(const_procinfo_ref);
+
+            //known A we can compute the deformation gradient analytically as F = A + I and from that the analytical value of the strain
+            Matrix F = IdentityMatrix(3,3) + A;
+            Matrix strain_mat = 0.5*(prod(trans(F),F)-IdentityMatrix(3,3));
+            std::vector<double> reference_strain = {strain_mat(0,0), strain_mat(1,1), 2.0*strain_mat(0,1)};
+
+            //also the stress can be computed analytically as
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
            // Also the stress can be computed analytically as
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                r_node.Coordinates() = r_node.GetInitialPosition() + r_node.GetSolutionStepValue(DISPLACEMENT); //here i update the node coordinates
+            }
+
+            p_element->InitializeSolutionStep(const_procinfo_ref);
+            p_element->InitializeNonLinearIteration(const_procinfo_ref);
+            p_element->CalculateLocalSystem(lhs,rhs,const_procinfo_ref);
+            p_element->FinalizeNonLinearIteration(const_procinfo_ref);
+            p_element->FinalizeSolutionStep(const_procinfo_ref);
+
+            //known A we can compute the deformation gradient analytically as F = A + I and from that the analytical value of the strain
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
            // Known A we can compute the deformation gradient analytically as F = A + I and from that the analytical value of the strain
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            r_node.AddDof(DISPLACEMENT_Y);
+            r_node.AddDof(DISPLACEMENT_Z);
+        }
+
+        std::vector<ModelPart::IndexType> element_nodes {1,2,3};
+        auto p_element = r_model_part.CreateNewElement("TotalLagrangianElement2D3N", 1, element_nodes, p_elem_prop);
+
+        p_element->Initialize(r_model_part.GetProcessInfo());
+
+        //define a matrix A and impose that the displacement on each node is u = A*x0
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
        // Define a matrix A and impose that the displacement on each node is u = A*x0
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+        // Set the element properties
+        auto p_elem_prop = r_model_part.CreateNewProperties(0);
+        p_elem_prop->SetValue(YOUNG_MODULUS, 2.0e+06);
+        p_elem_prop->SetValue(POISSON_RATIO, 0.3);
+        p_elem_prop->SetValue(THICKNESS, 0.01);
+        const auto &r_clone_cl = KratosComponents<ConstitutiveLaw>::Get("LinearElasticPlaneStress2DLaw");
+        p_elem_prop->SetValue(CONSTITUTIVE_LAW, r_clone_cl.Clone());
+
+        //constants for the computation of the stress
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
        // Constants for the computation of the stress
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            reference_stress[1] = c2 * reference_strain[0] + c1 * reference_strain[1];
+            reference_stress[2] = c3 * reference_strain[2];
+
+            std::vector<Vector> output_strains(1);
+            p_element->CalculateOnIntegrationPoints(GREEN_LAGRANGE_STRAIN_VECTOR,output_strains, r_model_part.GetProcessInfo());
+
+            std::vector<Vector> output_stress(1);
+            p_element->CalculateOnIntegrationPoints(PK2_STRESS_VECTOR,output_stress, r_model_part.GetProcessInfo());
+            
+            for(unsigned int i=0; i<reference_strain.size(); ++i)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
            KRATOS_CHECK_VECTOR_EQUAL(output_strains[0], reference_strain);
            KRATOS_CHECK_VECTOR_EQUAL(output_stress[0], reference_stress);
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                KRATOS_CHECK_DOUBLE_EQUAL(output_strains[0][i], reference_strain[i]);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                KRATOS_CHECK_DOUBLE_EQUAL(output_stress[0][i], reference_stress[i]);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion 
```

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            }
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_total_lagrangian.cpp]:
```suggestion
```

**Review by loumalouomega — APPROVED** [review]

**Review by philbucher — APPROVED** [review]

**Comment by RiccardoRossi** [comment]
> > **Description**
> > a simple test in cpp to verify that the output of strain and stress is ok for the TL element
> > Please mark the PR with appropriate tags:
> > ```
> > * FastPR
> > ```
> > 
> > 
> > **Changelog**
> 
> But we have the python tests, this is basically the same, I am Ok with it, but I don't understand why

It is true, however i have been debugging with this since we are having some propblems with the value output.
Having it in cpp is quite useful to this purpose and it does no harm...

**Pull request merged**